html_css/v2/main-course/transforms-animation.md

```diff
@@ -0,0 +1,176 @@
+
+### Optimization
+
+Optimization is a subject that can lead down quite a few rabbit-holes and is generally not worth focusing on at this intermediate stage. However there are a couple of things you need to keep in mind regarding CSS transitions and animations in general. Feel free to do your own research but if you've started to read about GPU threads regarding this particular topic then you've gone too far, although it is certainly interesting!
+
+We will start with "stacking contexts". Basically, these are formed when certain element scenarios are in place. A relevant scenario for us would be to transition a `transform` property like below:
+
+~~~css
+div {
+  width: 100px;
+  height: 100px;
+  transition: transform 2s 1s; 
+}
+
+div:hover {
+  transform: rotate(180deg);
+}
+~~~
+
+This has created a stacking context. If we were to make a bunch more stacking contexts through various other means (see [this article](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) for definitions) then, when it comes to rendering our initial transform, we would repaint not only our `div` element but also **every element that is stacked on top of it in the stack context**. If left unchecked, this can cause your once buttery-smooth transition to become slow and rough. This article from [Dzhavat Ushev](https://dzhavat.github.io/2021/02/18/debugging-layout-repaint-issues-triggered-by-css-transition.html) is a great example of catching and debugging a repaint issue in the wild.
+
+Also, you should keep your animations to only affecting `opacity` and `transform` if you want absolute best performance for animations on your web page. Yes, our first example above only carried out a simple `background-color` change, but even that was an expensive operation in itself. Have a read of this web.dev article for some handy hints on how to [create high-performance CSS animations](https://web.dev/animations-guide/). It also has a link to [CSS Triggers](https://csstriggers.com/) which will help you see how a CSS property will affect other elements on the page if it is animated. Look at the difference between the `background-color` and `transform` properties.
+
+Like we said earlier, optimization is a rabbit-hole so don't get too bogged down in the details here. What's important is that you have a solid understanding of these concepts and can apply them when necessary because hey, if you need to turn a button into a rainbow when it's hovered over, you better be able to transition that rainbow!
```

Comment:
as mentioned above.. I think we can remove the optimization rabbit hole statement here :D